File: scripts/update-basketball-data.py
#!/usr/bin/env python3
"""
Kentucky Basketball Data Auto-Updater (Simplified)

This script updates:
1. Total season record (W-L) for 2025-2026 season

You will manually update: Conference standings, KenPom, NET Rankings, Bracketology

Usage:
    python update-basketball-data.py
"""

import json
import os
import sys
from datetime import datetime
from typing import Dict, Optional
import cbbd
from cbbd.rest import ApiException

# Configuration
API_KEY = os.environ.get('BASKETBALL_API_KEY', '')
KENTUCKY_TEAM = 'Kentucky'
SEASON = 2026  # 2025-2026 season

# File paths
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
DATA_DIR = os.path.join(os.path.dirname(SCRIPT_DIR), 'data')
UPDATE_JSON_PATH = os.path.join(DATA_DIR, 'update.json')


def get_api_configuration():
    """Create and return the CBBD API configuration."""
    configuration = cbbd.Configuration(
        host="https://api.collegebasketballdata.com"
    )
    
    if API_KEY:
        configuration.access_token = API_KEY
    else:
        print("⚠️  Warning: No API key found. Set BASKETBALL_API_KEY environment variable.")
    
    return configuration


def fetch_team_record() -> Optional[Dict]:
    """Fetch Kentucky's current season record (W-L)."""
    print("Fetching Kentucky season record...")
    
    try:
        configuration = get_api_configuration()
        
        with cbbd.ApiClient(configuration) as api_client:
            stats_api = cbbd.StatsApi(api_client)
            
            try:
                stats_response = stats_api.get_team_season_stats(
                    season=SEASON,
                    team=KENTUCKY_TEAM
                )
                
                if stats_response and len(stats_response) > 0:
                    stat = stats_response[0]
                    wins = getattr(stat, 'wins', 0)
                    losses = getattr(stat, 'losses', 0)
                    
                    print(f"  ✓ Overall Record: {wins}-{losses}")
                    
                    return {
                        'overall_record': f"{wins}-{losses}",
                        'wins': wins,
                        'losses': losses
                    }
                else:
                    print("  ⚠️  No stats found for current season")
                    return None
                
            except ApiException as e:
                print(f"  ✗ Stats API error: {e}")
                return None
                
    except Exception as e:
        print(f"  ✗ Error fetching team record: {e}")
        return None


def update_json_file(record_data: Dict) -> bool:
    """Update the update.json file with record only."""
    print("Updating update.json file...")
    
    try:
        # Read existing data
        with open(UPDATE_JSON_PATH, 'r') as f:
            existing_data = json.load(f)
        
        # Ensure 2025 section exists (this represents 2025-2026 season)
        if '2025' not in existing_data:
            existing_data['2025'] = {'stats': {}, 'rankings': {}}
        
        # Update overall record
        if record_data:
            existing_data['2025']['rankings']['Overall Record'] = record_data['overall_record']
            print(f"  ✓ Updated Overall Record: {record_data['overall_record']}")
        
        # Write updated data back
        with open(UPDATE_JSON_PATH, 'w') as f:
            json.dump(existing_data, f, indent=4)
        
        print("  ✓ File updated successfully!")
        return True
        
    except FileNotFoundError:
        print(f"  ✗ Error: File not found: {UPDATE_JSON_PATH}")
        print(f"     Please make sure {UPDATE_JSON_PATH} exists")
        return False
    except json.JSONDecodeError as e:
        print(f"  ✗ Error: Invalid JSON in file: {e}")
        return False
    except Exception as e:
        print(f"  ✗ Error updating file: {e}")
        import traceback
        traceback.print_exc()
        return False


def main():
    """Main execution function."""
    print("=" * 70)
    print("Kentucky Basketball Auto-Updater - Simplified")
    print(f"Season: 2025-2026 (API season {SEASON})")
    print(f"Started at: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
    print("=" * 70)
    print()
    
    # Ensure data directory exists
    os.makedirs(DATA_DIR, exist_ok=True)
    
    success = True
    
    # Fetch team record
    record_data = fetch_team_record()
    if not record_data:
        print("⚠️  Warning: Could not fetch team record")
        success = False
    
    print()
    
    # Update the JSON file
    if record_data:
        if not update_json_file(record_data):
            success = False
    else:
        print("✗ No data to update")
        success = False
    
    print()
    print("=" * 70)
    print(f"Finished at: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
    
    if success:
        print("Status: SUCCESS ✓")
        print()
        print("Updated:")
        if record_data:
            print(f"  • Overall Record: {record_data['overall_record']}")
        print()
        print("You can manually update:")
        print("  • Conference Standing")
        print("  • KenPom Rankings")
        print("  • NET Rankings")
        print("  • Bracketology")
        print("=" * 70)
        return 0
    else:
        print("Status: FAILED ✗")
        print("=" * 70)
        return 1


if __name__ == '__main__':
    sys.exit(main())

File: scripts/update-basketball-data.js
#!/usr/bin/env node

/**
 * Kentucky Basketball Data Auto-Updater
 * 
 * This script automatically updates:
 * 1. Team stats and rankings in update.json
 * 2. Game results in 2025-schedule.json
 * 
 * Usage:
 * - Run manually: node update-basketball-data.js
 * - Schedule with cron: 0 */6 * * * /path/to/update-basketball-data.js
 */

const https = require('https');
const fs = require('fs');
const path = require('path');

// Configuration
const CONFIG = {
  apiKey: '0/5PdgRvOqvcUo9VqUAcXFUEYqXxU3T26cGqt9c6FFArBcyqE4BD3njMuwOnQz+3',
  apiBaseUrl: 'api.collegebasketballdata.com',
  kentuckyTeamId: 'KENT', // You may need to verify this ID
  dataDir: path.join(__dirname, 'data'),
  updateJsonPath: path.join(__dirname, 'data', 'update.json'),
  scheduleJsonPath: path.join(__dirname, 'data', '2025-schedule.json'),
  season: '2025'
};

/**
 * Make HTTPS request to the API
 */
function makeApiRequest(endpoint) {
  return new Promise((resolve, reject) => {
    const options = {
      hostname: CONFIG.apiBaseUrl,
      path: endpoint,
      method: 'GET',
      headers: {
        'Authorization': `Bearer ${CONFIG.apiKey}`,
        'Content-Type': 'application/json'
      }
    };

    const req = https.request(options, (res) => {
      let data = '';

      res.on('data', (chunk) => {
        data += chunk;
      });

      res.on('end', () => {
        if (res.statusCode === 200) {
          try {
            resolve(JSON.parse(data));
          } catch (e) {
            reject(new Error('Failed to parse JSON response'));
          }
        } else {
          reject(new Error(`API returned status code ${res.statusCode}: ${data}`));
        }
      });
    });

    req.on('error', (error) => {
      reject(error);
    });

    req.end();
  });
}

/**
 * Fetch Kentucky team stats and rankings
 */
async function fetchTeamStats() {
  console.log('Fetching Kentucky team stats...');
  
  try {
    // Try different common endpoints
    const endpoints = [
      `/teams/${CONFIG.kentuckyTeamId}/stats?season=${CONFIG.season}`,
      `/v1/teams/${CONFIG.kentuckyTeamId}/stats?season=${CONFIG.season}`,
      `/api/teams/${CONFIG.kentuckyTeamId}?season=${CONFIG.season}`,
      `/teams/stats?team=${CONFIG.kentuckyTeamId}&season=${CONFIG.season}`
    ];

    for (const endpoint of endpoints) {
      try {
        console.log(`Trying endpoint: ${endpoint}`);
        const data = await makeApiRequest(endpoint);
        return data;
      } catch (error) {
        console.log(`Endpoint failed: ${endpoint} - ${error.message}`);
      }
    }

    throw new Error('All endpoints failed');
  } catch (error) {
    console.error('Error fetching team stats:', error.message);
    return null;
  }
}

/**
 * Fetch team schedule and results
 */
async function fetchSchedule() {
  console.log('Fetching Kentucky schedule...');
  
  try {
    const endpoints = [
      `/teams/${CONFIG.kentuckyTeamId}/schedule?season=${CONFIG.season}`,
      `/v1/teams/${CONFIG.kentuckyTeamId}/schedule?season=${CONFIG.season}`,
      `/api/schedule?team=${CONFIG.kentuckyTeamId}&season=${CONFIG.season}`,
      `/schedule/${CONFIG.kentuckyTeamId}?season=${CONFIG.season}`
    ];

    for (const endpoint of endpoints) {
      try {
        console.log(`Trying endpoint: ${endpoint}`);
        const data = await makeApiRequest(endpoint);
        return data;
      } catch (error) {
        console.log(`Endpoint failed: ${endpoint} - ${error.message}`);
      }
    }

    throw new Error('All endpoints failed');
  } catch (error) {
    console.error('Error fetching schedule:', error.message);
    return null;
  }
}

/**
 * Update the update.json file with new stats
 */
function updateStatsFile(statsData) {
  console.log('Updating stats file...');
  
  try {
    // Read existing data
    const existingData = JSON.parse(fs.readFileSync(CONFIG.updateJsonPath, 'utf8'));
    
    // Update 2025 stats (you'll need to map the API response to your format)
    // This is a template - adjust based on actual API response structure
    if (statsData && statsData.stats) {
      existingData['2025'].stats = {
        "Offensive Rating": { 
          "value": statsData.stats.offensiveRating || existingData['2025'].stats['Offensive Rating'].value,
          "rank": statsData.stats.offensiveRatingRank || existingData['2025'].stats['Offensive Rating'].rank
        },
        "Defensive Rating": { 
          "value": statsData.stats.defensiveRating || existingData['2025'].stats['Defensive Rating'].value,
          "rank": statsData.stats.defensiveRatingRank || existingData['2025'].stats['Defensive Rating'].rank
        },
        // Add other stats mappings here
        ...existingData['2025'].stats
      };
    }

    if (statsData && statsData.rankings) {
      existingData['2025'].rankings = {
        "Overall Record": statsData.rankings.record || existingData['2025'].rankings['Overall Record'],
        "KenPom": statsData.rankings.kenpom || existingData['2025'].rankings.KenPom,
        "NET Rankings": statsData.rankings.net || existingData['2025'].rankings['NET Rankings'],
        // Add other ranking mappings here
        ...existingData['2025'].rankings
      };
    }

    // Write updated data
    fs.writeFileSync(CONFIG.updateJsonPath, JSON.stringify(existingData, null, 4));
    console.log('Stats file updated successfully!');
    return true;
  } catch (error) {
    console.error('Error updating stats file:', error.message);
    return false;
  }
}

/**
 * Update the 2025-schedule.json file with game results
 */
function updateScheduleFile(scheduleData) {
  console.log('Updating schedule file...');
  
  try {
    // Read existing schedule
    const existingSchedule = JSON.parse(fs.readFileSync(CONFIG.scheduleJsonPath, 'utf8'));
    
    if (!scheduleData || !Array.isArray(scheduleData.games)) {
      console.log('No schedule data to update');
      return false;
    }

    // Update each game with results from API
    for (const apiGame of scheduleData.games) {
      // Find matching game in schedule
      const scheduleGame = existingSchedule.find(game => {
        // Match by date and opponent
        const gameDate = new Date(game.date);
        const apiDate = new Date(apiGame.date);
        return gameDate.toDateString() === apiDate.toDateString();
      });

      if (scheduleGame && apiGame.status === 'completed') {
        // Game has been played, update result
        const kentuckyScore = apiGame.homeTeam === 'Kentucky' ? apiGame.homeScore : apiGame.awayScore;
        const opponentScore = apiGame.homeTeam === 'Kentucky' ? apiGame.awayScore : apiGame.homeScore;
        
        scheduleGame.result = kentuckyScore > opponentScore 
          ? `W ${kentuckyScore}-${opponentScore}`
          : `L ${kentuckyScore}-${opponentScore}`;
        
        console.log(`Updated: ${scheduleGame.date} - ${scheduleGame.result}`);
      }
    }

    // Write updated schedule
    fs.writeFileSync(CONFIG.scheduleJsonPath, JSON.stringify(existingSchedule, null, 4));
    console.log('Schedule file updated successfully!');
    return true;
  } catch (error) {
    console.error('Error updating schedule file:', error.message);
    return false;
  }
}

/**
 * Main execution function
 */
async function main() {
  console.log('=== Kentucky Basketball Data Updater ===');
  console.log(`Started at: ${new Date().toLocaleString()}`);
  console.log('');

  // Ensure data directory exists
  if (!fs.existsSync(CONFIG.dataDir)) {
    fs.mkdirSync(CONFIG.dataDir, { recursive: true });
  }

  try {
    // Fetch and update stats
    const statsData = await fetchTeamStats();
    if (statsData) {
      updateStatsFile(statsData);
    }

    // Fetch and update schedule
    const scheduleData = await fetchSchedule();
    if (scheduleData) {
      updateScheduleFile(scheduleData);
    }

    console.log('');
    console.log('=== Update Complete ===');
    console.log(`Finished at: ${new Date().toLocaleString()}`);
  } catch (error) {
    console.error('Fatal error:', error.message);
    process.exit(1);
  }
}

// Run the script
if (require.main === module) {
  main();
}

module.exports = { main, fetchTeamStats, fetchSchedule, updateStatsFile, updateScheduleFile };
